airbyte-cdk/python/airbyte_cdk/sources/message/repository.py

```diff
@@ -15,14 +15,43 @@ class MessageRepository(ABC):
     def emit_message(self, message: AirbyteMessage) -> None:
         raise NotImplementedError()
 
+    @abstractmethod
+    def log_message(self, level: Level, message_provider: Callable[[], AirbyteMessage]) -> None:
```

Comment:
FIXME: should be `Callable[[], str]` and as we would support any JSON, I'll probably update it to 
`JsonType = Union[dict[str, "JsonType"], list["JsonType"], str, int, float, bool, None]`